truetype/truetype.go

```diff
@@ -306,6 +339,142 @@ func (f *Font) parseMaxp() error {
+
+		switch nameID {
+		case NameIDFontFamily, NameIDFontSubfamily:
+			if u16(f.name, offset+8) == 0 { // Offset + 8 is the location of the string length.  If zero, skip it.
+				break
+			}
+			switch platformID {
+			case PLATFORM_UNICODE:
+				foundUnicode[nameID-1] = offset
+			case PLATFORM_MACINTOSH:
+				if languageID == 0 { // The language ID for English is zero.
+					foundAppleEnglish[nameID-1] = offset
+				} else if encodingID == 0 { // A zero encoding ID means Roman.
+					foundAppleRoman[nameID-1] = offset
+				}
+			case PLATFORM_MICROSOFT:
+				if foundMicrosoft[nameID-1] == 0 || languageID&0x3FF == 0x009 { // Prefer Microsoft fonts in English.
+					if encodingID == MICROSOFT_SYMBOL_CS || encodingID == MICROSOFT_UNICODE_CS || encodingID == MICROSOFT_UCS_4 {
+						foundMicrosoftEnglish = languageID&0x3FF == 9
+						foundMicrosoft[nameID-1] = offset
+					}
+				}
+			}
+		}
+	}
+
+	foundApple = foundAppleRoman
+	if foundAppleEnglish[0] > 0 {
+		foundApple = foundAppleEnglish
+	}
+
+	var nameOffset, nameLength, styleOffset, styleLength uint16
+	nameRecord := &nameInfo{}
+	// Retrieve ASCII version of the values.
+	switch {
+	case foundMicrosoft[0] > 0 && !(foundApple[0] > 0 && !foundMicrosoftEnglish):
+		nameOffset, styleOffset = u16(f.name, foundMicrosoft[0]+10)+stringOffset, u16(f.name, foundMicrosoft[1]+10)+stringOffset
+		nameLength, styleLength = u16(f.name, foundMicrosoft[0]+8), u16(f.name, foundMicrosoft[1]+8)
+	case foundApple[0] > 0:
+		nameOffset, styleOffset = u16(f.name, foundApple[0]+10)+stringOffset, u16(f.name, foundApple[1]+10)+stringOffset
+		nameLength, styleLength = u16(f.name, foundApple[0]+8), u16(f.name, foundApple[1]+8)
+	case foundUnicode[0] > 0:
+		nameOffset, styleOffset = u16(f.name, foundUnicode[0]+10)+stringOffset, u16(f.name, foundUnicode[1]+10)+stringOffset
+		nameLength, styleLength = u16(f.name, foundUnicode[0]+8), u16(f.name, foundUnicode[1]+8)
+	}
+
+	nameRecord.values[NameIDFontFamily] = f.nameEntryInASCII(f.name[nameOffset:nameOffset+nameLength], true)
+	nameRecord.values[NameIDFontSubfamily] = f.nameEntryInASCII(f.name[styleOffset:styleOffset+styleLength], true)
+
+	f.nameRecord = nameRecord
+
+	return nil
+}
+
+func (f *Font) nameEntryInASCII(b []byte, utf16 bool) string {
+	var buf []byte
+	if utf16 { // Equivalent to tt_name_ascii_from_utf16.
+		j := len(b)
+		if j&1 == 1 {
+			return ""
+		}
+
+		for i := 0; i < j; i += 2 {
+			el := u16(b, i)
+			if el == 0 {
+				continue
+			} else if el < 32 || el > 127 {
+				buf = append(buf, '?')
+			} else {
+				buf = append(buf, byte(el))
+			}
+		}
+	} else { // Equivalent to tt_name_ascii_from_other.
+		for _, el := range b {
+			if el == 0 {
+				continue
+			} else if el < 32 || el > 127 {
+				buf = append(buf, '?')
+			} else {
+				buf = append(buf, el)
+			}
+		}
+	}
+
+	return string(buf)
+}
+
+// Name returns the font's name.
+func (f *Font) Name(id NameID) string {
+	if f.nameRecord == nil {
```

Comment:
This whole function should just be
return f.parseName(id)
but then, just move parseName's code into here.